software/displayerang.livecodescript

```diff
@@ -0,0 +1,1402 @@
+               put true into tAgain
+               delete item itemOffset(tStart,pData["layers"][tStartLayer]) of pData["layers"][tStartLayer]
+               put tStart & "," after pData["layers"][tStartLayer -1]
+               put tStartLayer -1 into pData["nodes"][tStart]["layer"]
+               put cr & tStart & "-" after tShuffles
+               filter tShuffles without empty
+            else
+               -- shuffle end more positive
+               put true into tAgain
+               delete item itemOffset(tEnd,pData["layers"][tEndLayer]) of pData["layers"][tEndLayer]
+               put tEnd & "," after pData["layers"][tEndLayer +1]
+               put tEndLayer +1 into pData["nodes"][tEnd]["layer"]
+               put cr & tEnd & "+" after tShuffles
+               filter tShuffles without empty
+            end if
+         end if -- is tStart < tEnd?
+      end repeat -- tLink in sLinksVisited
+   end repeat -- tAgain
+   -- order in layer by node #, ascending ----------------------------------------------------------
+   repeat for each key tLayer in pData["layers"]
+      sort items of pData["layers"][tLayer] ascending numeric
+      if char 1 of pData["layers"][tLayer] is "," then delete char 1 of pData["layers"][tLayer]
+   end repeat
+   -- ! deal with elements that are visible but weren't discovered by middleOut
+   repeat for each word tKind in "nodes links"
+      repeat for each line tRec in the keys of pData[tKind]
+         if tRec is not among the items of pData["visited"][tKind] then
+            put tRec into item (the number of items of pData["unvisited"][tKind])+1 of pData["unvisited"][tKind]
+         end if
+      end repeat
+   end repeat
+   put pData into pDraw
+end middleOut
+
+on assignLayers @pAll,pSeed,pTheLayer -- ! the recursive part of middleOut
+   put "on assignLayers" into xX["write"];    devLog xX
+   -- pAll = pData (just a copy of sGraphData at the moment); this adds display info to it
+   -- pSeed = the node to find neighbors of
+   -- pTheLayer = the layer pSeed was assigned before this function was called
+   --- list all next & prev links
+   --- format as links
+   --- don't add if already on list of visited links
+   --- add node to layer; prev -1; next +1
+   --- call this function again on this node
+   --- terminal case is when this node has no unvisited links
+   if pSeed is not among the items of pAll["visited"]["nodes"] then
+      put pSeed & "," after pAll["visited"]["nodes"]
+      put pSeed & "," after pAll["layers"][pTheLayer]
+      put pTheLayer into pAll["nodes"][pSeed]["layer"]
+      repeat for each word tDir in "next prev"
+         --         repeat for each item tLink in pAll["nodes"][pSeed][tDir]
+         repeat for each line tLink in pAll["nodes"][pSeed][tDir]
+            if tLink is not among the items of pAll["visited"]["links"] then
+               put tLink & "," after pAll["visited"]["links"]
+               filter pAll["visited"]["links"] without empty
```

Comment:
@matthewmaier I can't see what the point of filtering without empty is here as tLink is a line and you are adding tLink & comma it seems impossible to end up with an empty line.